app/controllers/programs/sites_controller.rb

```diff
@@ -1,10 +1,77 @@
 class Programs::SitesController < SitesController
   before_action :set_site_program
+  before_action :set_site, only: %i[ edit destroy ]
+  before_action :set_units
+
+  # GET /sites/new
+  def new
+    @site = Site.new
+    authorize @site
+  end
+
+  # GET /sites/1/edit
+  def edit
+    session[:return_to] = request.referer
+  end
+
+  def edit_program_sites
+    @site = Site.new
+    authorize Site
+  end
+
+  # POST /sites or /sites.json
+  def create
+    if params[:site_id].present?
+      @site = Site.find(params[:site_id])
+      authorize @site
+      if @site_program.sites << @site
+        @site = Site.new
+        flash.now[:notice] = "The site was added"
+      end
+    else
+     @site = Site.new(site_params)
+     authorize @site
+      if @site.save
+        @site_program.sites << @site
+        @site = Site.new
+        flash.now[:notice] = "The site was added"
+      end
+    end
+    @sites = @site_program.sites
+    @all_sites = Site.all - @sites
+  end
+
+  # DELETE /sites/1 or /sites/1.json
+  def destroy
+    # remove site from program, don't destroy
+    if @site_program.sites.delete(@site)
+      flash.now[:notice] = "The site was removed from the program"
+    end
+    @site = Site.new
+    @sites = @site_program.sites
+    @all_sites = Site.all - @sites
+  end
 
   private
 
-  def set_site_program
+    def set_site_program
```

Comment:
@rsmoke  No, I can't. set_site_program runs for every action, set_site only for [ edit remove_site_from_program ]

before_action :set_site_program
before_action :set_site, only: %i[ edit remove_site_from_program ]
site_program should exits for all methods (including new and create) because we add a new site to a specific program. We don't create a site without a program.